### DISCUSSION TITLE
Is it necessary to touch a %ghost?

### DISCUSSION THREAD

**Discussion by hroncok** [rpm-software-management/rpm#3380] [Q&A] [answered]
A very very old documentation says:

> There are times when a file should be owned by the package but not installed - log files and state files are good examples of cases you might desire this to happen.
> 
> The way to achieve this, is to use the %ghost directive. By adding this directive to the line containing a file, RPM will know about the ghosted file, but will not add it to the package. **However it still needs to be in the buildroot.** Here's an example of %ghost in action. 
>
> ```
> %install
> touch $RPM_BUILD_ROOT%{_localstatedir}/log/blather.log
> …
> %files
> …
> %ghost %{_localstatedir}/log/blather.log
> ```

http://ftp.rpm.org/max-rpm/s1-rpm-inside-files-list-directives.html#S3-RPM-INSIDE-FLIST-GHOST-DIRECTIVE

The current docs say no such thing:

https://rpm-software-management.github.io/rpm/manual/spec.html#ghost

It is possible to build a package with `%ghost %{_localstatedir}/log/blather.log` without first touching the path (at least with RPM 4.20), but rpmlint 2.5.0 says:

```
joft.noarch: E: non-readable /var/log/blather.log 0
```

That brings me to this question:

 - Is it necessary to create (e.g. touch) `%ghost` files in the buildroot, or was this only true in the past?
 - If it is necessary, should RPM fail to build the package when the %ghost is not in the buildroot?
 - If it is not necessary, is this a bug in rpmlint?
 

![a ghost cartoon](https://github.com/user-attachments/assets/a734e818-4b44-4d62-9a16-1f6fbdfa7700)

**Comment by pmatilai** [comment, marked as answer]
It's a bug in rpmlint. 

Rpm hasn't required %ghost files to exist in the buildroot since 2009 (07524b67b6cfe501bdae3df45c212b7c1c129da4) and since 2012 has supported %ghost directories in similar manner too (07524b67b6cfe501bdae3df45c212b7c1c129da4)

**Reply by hroncok** [reply]
Reported https://github.com/rpm-software-management/rpmlint/issues/1287

**Reply by pemensik** [reply]
In that case we need to update [Alternatives](https://docs.fedoraproject.org/en-US/packaging-guidelines/Alternatives/) in  Fedora packaging guidelines. Because it is still used in examples, including *the original locations MUST be touched* at [How to use alternatives](https://docs.fedoraproject.org/en-US/packaging-guidelines/Alternatives/#_how_to_use_alternatives)

**Reply by pemensik** [reply]
Created issue for it at: https://forge.fedoraproject.org/packaging/guidelines/issues/1554. Have not enough free time to make a PR now.

**Comment by ffesti** [comment]
You still might want to create the ghost file - e.g. if you want to reserve the space during installation. The meta data is taken from the file if it exists.

**Comment by pemensik** [comment]
But in that case, ghost files are listed in ``rpm -qlv setup`` with file permissions and file mode. If ghost file is not touched first, what attributes and owner will it have? If it should not be visible in ``rpm -V setup`` if file mode changes, it should be given some defaults. That is why I recently used ``install -m 0640 /dev/null %{buildroot}/...`` instead of touch + chmod.

**Reply by pmatilai** [reply]
A %ghost that does not exist in the buildroot needs an  explicit %attr() to tell rpm those things.

**Reply by pmatilai** [reply]
...or so I thought. I could've sworn we even had an error message for that case, but...

**Reply by pmatilai** [reply]
Apparently it just picks up whatever the %defattr() at that point is.

**Reply by pmatilai** [reply]
And okay, I haven't completely lost my marbles. We *did* have such a behavior and error message, but it was changed 14 years ago in d6775a746bbaf8ddc8a91c9fae68ae91af1cc570. How the time flies!

**Reply by pemensik** [reply]
Hmm, but there is another glitch. If you switch also man pages using alternatives with undefined compression suffix, then ``%ghost example.1*`` will do unexpected thing. Because that file does not exist and was not touched, you need to specify also man page suffix explicitly. If you do not handle it explicitly, the star will remain unexpanded in current rawhide. Which is very likely not what you wanted.

Shouldn't it then default to umask defined permissions? %defattr should be optional. But clearly it is not, because I had ``--------`` permissions on files without explicit %attr. Is that a good default?

I think we should provide macro for default man page suffix in Fedora, like OpenSUSE has for a long time.
